### PR TITLE
Md/vk

### PR DESCRIPTION
**Pull request by la10736**
Implements [ZKV-230](https://horizenlabs.atlassian.net/browse/ZKV-230)

Introduce the exstrinsic for registring Vk and add an optional (for retro campatibility) variant `VkOrHash` to submit the proof. 

Also add E2E tests and refactor athe E2E tests little bit.

### CODE REVIEW

File: e2e-tests/README.md [hunk 1]
```diff
@@ -34,54 +37,71 @@ Each end-to-end test requires the following files:
...
 ### Execute the test suite entirely
 
 Run from within the `e2e-tests` directory:
+
 - `yarn install`
 - `yarn test`
 
-The last command invokes the script `test_runner.sh`.
+The last command invokes the script `test_runner.sh`. If you want to use the binary compiled in
+debug flavor you can use `--debug` switch. Follow line execute all tests running the node binary
```
Comment (DanieleDiBenedetto) [inline e2e-tests/README.md:56]:
```suggestion
debug flavor you can use `--debug` switch. The following line executes all the tests running the node binary
```

File: e2e-tests/README.md [hunk 2]
```diff
@@ -34,54 +37,71 @@ Each end-to-end test requires the following files:
...
+from `target/debug` folder.
+
+```sh
+yarn test --debug
+```
 
 ### Execute a single test
 
-The following instructions are for the local execution environment only, without taking into account Kubernetes or Podman - supported by zombienet, but not mandatory.
+You can run a list of tests simply by indicate the names while running tests:
```
Comment (DanieleDiBenedetto) [inline e2e-tests/README.md:65]:
```suggestion
You can run a list of tests by simply specifying the names while running the tests:
```

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]
Minor changes. Not a fan of the default vk in case not specified though

File: pallets/settlement-fflonk/src/lib.rs [hunk 3]
```diff
@@ -99,29 +106,102 @@ pub mod pallet {
...
+                .try_into()
+                .map_err(|_| Error::<T>::InvalidVerificationKey)?;
+            let hash = hash_vk(&vk);
+            Vks::<T>::insert(hash, vk);
+            Self::deposit_event(Event::VkRegistered { hash });
+            Ok(().into())
+        }
+    }
+
+    const PREFIX: &[u8; 6] = b"fflonk";
```
[the remark this answers is not in the document]
Comment (la10736) [reply pallets/settlement-fflonk/src/lib.rs:170]: No, take a look to line 181 and 178 that add this character correctly.

**Review by la10736 — COMMENTED** [review]
> Minor changes. Not a fan of the default vk in case not specified though 

Me too, but I've used this trick to make a smoother transaction... I'll remove it when I'll abstract the verifier.

**Review by DanieleDiBenedetto — APPROVED** [review]

**Review by drgora — APPROVED** [review]
Nice refactoring for e2e tests, thanks!
This looks good to me, I just left two comments/clarification questions, but nothing blocking

File: pallets/settlement-fflonk/src/lib.rs [hunk 4]
```diff
@@ -99,29 +106,102 @@ pub mod pallet {
...
+                    T::OnProofVerified::on_proof_verified(compute_fflonk_hash(
+                        *raw_proof, vk_or_hash,
+                    ))
+                })
                 .map(Into::into)
                 .map_err(Into::into)
         }
+
+        #[pallet::call_index(1)]
+        pub fn register_vk(_origin: OriginFor<T>, vk: Box<Vk>) -> DispatchResultWithPostInfo {
```
[the remark this answers is not in the document]
Comment (la10736) [reply pallets/settlement-fflonk/src/lib.rs:158]: Yes we had.

We're inheriting it from ` #[pallet::call(weight(<T as Config>::WeightInfo))]` impl's attribute: if the weight trait have a method that match with our extrinsic it uses this method.

File: e2e-tests/js_scripts/0007-proof_with_vk.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    if (receivedEvents(await submitProof(api.tx.settlementFFlonkPallet, alice, VALID_PROOF, { Hash: VKEY_HASH }))) {
+        return ReturnCode.ErrAcceptAnUnregisteredHash;
+    };
+
+    events = await registerVk(api.tx.settlementFFlonkPallet, alice, vk);
+    if (!receivedEvents(events)) {
+        return ReturnCode.ErrVkRegistrationFailed;
+    };
+    const vkHash = events[0].data[0];
+
```
[the remark this answers is not in the document]
Comment (la10736) [reply e2e-tests/js_scripts/0007-proof_with_vk.js:79]: Right, I'll fix it! 
THX!

**Pull request merged**